sentry_streams/sentry_streams/examples/batching.py

```diff
@@ -1,32 +1,33 @@
-import json
-
-from sentry_streams.pipeline import Batch, FlatMap, Map, streaming_source
-from sentry_streams.pipeline.batch import unbatch
-from sentry_streams.pipeline.function_template import InputType
-
-
-def build_batch_str(batch: list[InputType]) -> str:
-    d = {"batch": batch}
-
-    return json.dumps(d)
+from typing import Callable, MutableSequence, Union, cast
 
+from sentry_kafka_schemas.schema_types.ingest_metrics_v1 import IngestMetric
 
-def build_message_str(message: str) -> str:
-    d = {"message": message}
-
-    return json.dumps(d)
+from sentry_streams.pipeline import Batch, FlatMap, streaming_source
+from sentry_streams.pipeline.batch import unbatch
+from sentry_streams.pipeline.chain import Parser, Serializer
+from sentry_streams.pipeline.message import Message
+from sentry_streams.pipeline.msg_parser import json_parser, json_serializer
 
+pipeline = streaming_source(
+    name="myinput",
+    stream_name="ingest-metrics",
+)
 
-pipeline = (
-    streaming_source(
-        name="myinput",
-        stream_name="events",
-    )
-    .apply("mybatch", Batch(batch_size=5))  # User simply provides the batch size
-    .apply("myunbatch", FlatMap(function=unbatch))
-    .apply("mymap", Map(function=build_message_str))
-    .sink(
-        "kafkasink",
-        stream_name="transformed-events",
-    )  # flush the batches to the Sink
+# TODO: Figure out why the concrete type of InputType is not showing up in the type hint of chain1
+chain1 = pipeline.apply("parser", Parser(msg_type=IngestMetric, deserializer=json_parser)).apply(
+    "mybatch", Batch(batch_size=5)
+)  # User simply provides the batch size
+
+chain2 = chain1.apply(
+    "myunbatch",
+    FlatMap(
+        function=cast(
+            Union[Callable[[Message[MutableSequence[IngestMetric]]], Message[IngestMetric]], str],
```

Comment:
Because `unbatch` is a function with generic arguments, the user unfortunately has to explicitly cast it to the concrete type in their pipeline. This shouldn't be difficult because the incoming `ExtensibleChain` already has concrete type hints. 

Obviously this specific `cast()` can also be avoided if users write and use a custom unbatcher.

Reply:
I wonder whether it would be easier if we had  `Batch` and `Unbatch` appliers without having to go through the FlatMap.

Reply:
out of the scope for now.